### PR TITLE
Make GeoContext mapping idempotent

### LINKED ISSUE #8937: Completion type mapping is not idempotent
Completion mapping is not idempotent in the same way other field types are, probably due to the internal re-ordering of the `precision` array.

In the example below you can see that updating the mapping twice causes an error, if you start afresh and use the `precision` values `[ 1,2,3 ]` the issue goes away.

**Note** when you `PUT` the mapping as below and then `GET` it back you'll notice the ordering of the precision array is now `[2,3,1,4]`.

```bash
#!/bin/bash

################################################
# Completion Mapping Idempotence
################################################

ES='localhost:9200';

# drop/create index
curl -s -XDELETE "$ES/testcase?pretty=true" >/dev/null;
curl -s -XPUT "$ES/testcase?pretty=true" >/dev/null;

# update mapping
update_mapping() {
  curl -s -XPUT "$ES/testcase/_mapping/mytype?pretty=true" -d '
  {
     "properties": {
        "suggest": {
           "type": "completion",
           "context": {
              "location": {
                 "type": "geo",
                 "precision": [ 1, 2, 3, 4 ]
              }
           }
        }
     }
  }';
}

update_mapping;
# "acknowledged" : true

update_mapping;
# "error" : "MergeMappingException[Merge failed with failures {[mapper [suggest] has different 'context_mapping' values]}]"

curl -s "$ES/testcase/mytype/_mapping?pretty=true" | grep precision
# "precision" : [ 2, 3, 1, 4 ]
```

cc/ @markharwood as it may be related to https://github.com/elasticsearch/elasticsearch/commit/c0aef4adc4ffaf791ed9a42864cc578b7dc50ffc and @areek as it relates to context suggesters

### LINKED ISSUE #10581: Endless re-syncing mappings
We're running into an recurring issue of late where any change in cluster state (typically a node restart during an upgrade for example) results in the following errors streaming to the log of the currently elected master, as well as very high I/O on all eligible masters:

```
[2015-04-13 21:02:37,917][WARN ][cluster.metadata         ] [elasticsearch-master2.localdomain] [pelias] 
re-syncing mappings with cluster state for types [[osmway, locality, openaddresses, geonames
osmaddress, admin0, neighborhood, admin2, osmnode, admin1, local_admin]]
``` 

We are presently running 1.5.1, but we had the same issue on 1.5.0. I'm cannot say with certainty whether we experienced the issue under 1.4.x. The production cluster setup is as follows, but we have also seen this problem in local development (single node):

- 3 master eligible nodes
- 2 routing (no data, no master) nodes
- 8 data nodes
- 40 shards, no replicas, ~600gb of data

The problematic mapping seems to be the `location` part of the suggester field:

    PUT pelias
    {
      "mappings": {
        "admin0": {
          "properties": {
            "suggest": {
              "context": {
                "dataset": {
                  "default": [],
                  "path": "_type",
                  "type": "category"
                },
                "location": {
                  "default": [],
                  "neighbors": true,
                  "path": "center_point",
                  "precision": [2,3,1,5,4],
                  "type": "geo"
                }
              },
              "max_input_length": 50,
              "payloads": false,
              "preserve_position_increments": false,
              "preserve_separators": false,
              "type": "completion"
            }
          }
        }
      }
    }

### PR DESCRIPTION
**Pull request by areek**
closes #10581
closes #8937

### CODE REVIEW

File: src/test/java/org/elasticsearch/search/suggest/ContextSuggestSearchTests.java [hunk 1]
```diff
@@ -153,7 +155,26 @@ public void testMultiLevelGeo() throws Exception {
...
+                .field("type", "completion")
+                .startObject("context")
+                .startObject("location")
+                .field("type", "geo")
+                .array("precision", 1, 2, 3, 4)
+                .endObject()
+                .endObject().endObject()
+                .endObject().endObject();
+
+        assertAcked(prepareCreate(INDEX).setSettings(ImmutableSettings.builder().put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, 1)).addMapping(TYPE, mapping.string()));
```
Comment (rjernst) [inline src/test/java/org/elasticsearch/search/suggest/ContextSuggestSearchTests.java]: Why does this need a single shard?

File: src/test/java/org/elasticsearch/search/suggest/ContextSuggestSearchTests.java [hunk 2]
```diff
@@ -153,7 +155,26 @@ public void testMultiLevelGeo() throws Exception {
...
+
+    @Test
+    public void testMappingIdempotency() throws Exception {
+        XContentBuilder mapping = jsonBuilder().startObject().startObject(TYPE)
+                .startObject("properties").startObject("completion")
+                .field("type", "completion")
+                .startObject("context")
+                .startObject("location")
+                .field("type", "geo")
+                .array("precision", 1, 2, 3, 4)
```
Comment (rjernst) [inline src/test/java/org/elasticsearch/search/suggest/ContextSuggestSearchTests.java]: Should we pass in a random order?

**Comment by rjernst** [comment]
@areek The fix LGTM, just a couple questions about the test.

File: src/test/java/org/elasticsearch/search/suggest/ContextSuggestSearchTests.java (hunk 1 above)
Comment (areek) [inline src/test/java/org/elasticsearch/search/suggest/ContextSuggestSearchTests.java]: The bug could be reproduced using a single shard, changed it to test with multiple shards

File: src/test/java/org/elasticsearch/search/suggest/ContextSuggestSearchTests.java (hunk 2 above)
Comment (areek) [inline src/test/java/org/elasticsearch/search/suggest/ContextSuggestSearchTests.java]: Thanks for the suggestion, fixed

**Pull request closed without merging**